File: JANL/UserControls/DGVTemplateEditor.vb
Imports System.ComponentModel
Imports System.Drawing
Imports JANL.Types

Public Class DGVTemplateEditor
	Private ReadOnly DT As DataTable = New DataTable()
	Private Design As Design
	Private SkipDGVEvent As Boolean

	Public Function GetTemplate() As DGVTemplate
		Return Design.ToTemplate()
	End Function

	Public Sub SetTemplate(T As DGVTemplate)
		Design = New Design(T)
		ReCreateDGV()

		BS_Design.DataSource = Design
		BS_Columns.DataSource = Design.Columns
	End Sub

	Private Sub B_Export_Click(sender As Object, e As EventArgs) Handles B_Export.Click
		TB_XML.Text = DGVManager.ToXML(GetTemplate())
	End Sub

	Private Sub B_Import_Click(sender As Object, e As EventArgs) Handles B_Import.Click
		Try
			Dim T = DGVManager.FromXML(TB_XML.Text)
			SetTemplate(T)
		Catch ex As Exception
			Msgs.ShowError(ex.Message)
		End Try
	End Sub

	Private Sub CheckOrder()
		For Each C As DataGridViewTextBoxColumn In DGV_Template.Columns
			Debug.Print($"{C.HeaderText} {C.Index} {C.DisplayIndex}")
			If C.Index = C.DisplayIndex Then Continue For Else Exit For
			Exit Sub
		Next

		Dim TC = DGV_Template.Columns.Cast(Of DataGridViewTextBoxColumn).OrderBy(Function(C) C.DisplayIndex).Select(Function(C) DirectCast(C.Tag, DesignColumn).ToTemplate()).ToList()
		Dim T = New DGVTemplate With {.Columns = TC}
		SetTemplate(T)
	End Sub

	Private Sub DGVTemplateEditor_Load(sender As Object, e As EventArgs) Handles MyBase.Load
		Dim T = New DGVTemplate With {
			.Columns = New List(Of DGVTemplateColumn) From {
				New DGVTemplateColumn("Column1", "Column1 Name", 100, DataGridViewAutoSizeColumnMode.None, False),
				New DGVTemplateColumn("Column2", "Column2 Name", 200, DataGridViewAutoSizeColumnMode.None, True),
				New DGVTemplateColumn("Column3", "Column3 Name", 200, DataGridViewAutoSizeColumnMode.Fill, True),
				New DGVTemplateColumn("Column4", "Column4 Name", 300, DataGridViewAutoSizeColumnMode.NotSet, True)}}
		SetTemplate(T)
	End Sub

	Private Sub ReCreateDGV()
		'If SkipDGVEvent Then Return
		SkipDGVEvent = True

		DGV_Template.Columns.Clear()
		Design.Columns.ForEach(Sub(C) DGV_Template.Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = C.Name, .Tag = C}))

		SkipDGVEvent = False
		RefreshDGV()
	End Sub

	Private Sub RefreshDGV()
		'If SkipDGVEvent Then Return
		SkipDGVEvent = True

		For Each C As DataGridViewTextBoxColumn In DGV_Template.Columns
			Dim DC = DirectCast(C.Tag, DesignColumn)
			C.HeaderText = DC.Header
			C.AutoSizeMode = DC.AutoSizeMode
			C.Width = DC.Width
			C.DataPropertyName = DC.Name
			C.SortMode = DataGridViewColumnSortMode.NotSortable
			C.DefaultCellStyle.ForeColor = DC.ForeColor
			C.DefaultCellStyle.BackColor = If(DC.Visible, Color.White, Color.SlateGray)
			'C.DefaultCellStyle.Font = C.Font
		Next

		Dim DT = New DataTable()
		Design.Columns.ForEach(Sub(C) DT.Columns.Add(New DataColumn(C.Name)))
		Dim R = DT.NewRow()
		Design.Columns.ForEach(Sub(C) R(C.Name) = $"{C.Name} Data")
		DT.Rows.Add(R)
		DGV_Template.DataSource = DT

		If BS_Columns.Position > 0 Then
			DGV_Template.ClearSelection()
			DGV_Template.Item(BS_Columns.Position, 0).Selected = True
		End If

		SkipDGVEvent = False
	End Sub

#Region "BS Events"

	Private Sub BS_Columns_AddingNew(sender As Object, e As AddingNewEventArgs) Handles BS_Columns.AddingNew

	End Sub

	Private Sub BS_Columns_CurrentChanged(sender As Object, e As EventArgs) Handles BS_Columns.CurrentChanged
		PG_Column.SelectedObject = BS_Columns.Current
	End Sub

	Private Sub BS_Columns_CurrentItemChanged(sender As Object, e As EventArgs) Handles BS_Columns.CurrentItemChanged
		RefreshDGV()
	End Sub

	Private Sub BS_Columns_ListChanged(sender As Object, e As ListChangedEventArgs) Handles BS_Columns.ListChanged
		If (e.ListChangedType = ListChangedType.ItemDeleted OrElse e.ListChangedType = ListChangedType.ItemAdded) Then
			ReCreateDGV()
		End If
	End Sub

#End Region

#Region "DGV Events"

	Private Sub DGV_Template_CellMouseClick(sender As Object, e As DataGridViewCellMouseEventArgs) Handles DGV_Template.CellMouseClick
		If e.ColumnIndex >= 0 Then BS_Columns.Position = e.ColumnIndex
	End Sub

	Private Sub DGV_Template_CellMouseDown(sender As Object, e As DataGridViewCellMouseEventArgs) Handles DGV_Template.CellMouseDown
		Debug.Print($"Down: {e.Button.ToString}")
	End Sub

	Private Sub DGV_Template_CellMouseUp(sender As Object, e As DataGridViewCellMouseEventArgs) Handles DGV_Template.CellMouseUp
		Debug.Print($"Up: {e.Button.ToString}")
		If e.Button = MouseButtons.Left Then CheckOrder()
	End Sub

	Private Sub DGV_Template_ColumnDisplayIndexChanged(sender As Object, e As DataGridViewColumnEventArgs) Handles DGV_Template.ColumnDisplayIndexChanged
		Debug.Print($"Order I:{e.Column.HeaderText} DI:{e.Column.DisplayIndex}")
	End Sub

	Private Sub DGV_Template_ColumnHeaderMouseClick(sender As Object, e As DataGridViewCellMouseEventArgs) Handles DGV_Template.ColumnHeaderMouseClick
		If e.ColumnIndex >= 0 Then BS_Columns.Position = e.ColumnIndex
	End Sub

#End Region

End Class

File: JANL.WinForms/UserControls/DGVTemplateEditor.cs
using JANL.Types;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace JANL.UserControls
{
    public partial class DGVTemplateEditor : UserControl
    {
        private Design Design;
        private bool SkipDGVEvent;

        public DGVTemplate GetTemplate() => Design.ToTemplate();

        public void SetTemplate(DGVTemplate Template)
        {
            Design = new Design(Template);
            ReCreateDGV();

            BS_Design.DataSource = Design;
            BS_Columns.DataSource = Design.Columns;
        }

        private void B_Export_Click(object sender, EventArgs e) => TB_XML.Text = DGVManager.ToXML(GetTemplate());

        private void B_Import_Click(object sender, EventArgs e)
        {
            try
            {
                var T = DGVManager.FromXML(TB_XML.Text);
                SetTemplate(T);
            }
            catch (Exception ex)
            {
                Msgs.ShowError(ex.Message);
            }
        }

        private void CheckOrder()
        {
            foreach (DataGridViewTextBoxColumn C in DGV_Template.Columns)
            {
                Debug.Print($"{C.HeaderText} { C.Index} { C.DisplayIndex}");
                if (C.Index != C.DisplayIndex) { break; }
                return;
            }
            var TC = DGV_Template.Columns.Cast<DataGridViewTextBoxColumn>().OrderBy(C => C.DisplayIndex).Select(C => ((DesignColumn)C.Tag).ToTemplate()).ToList();
            var T = new DGVTemplate { Columns = TC };
            SetTemplate(T);
        }

        private void DGVTemplateEditor_Load(object sender, EventArgs e)
        {
            var TC = new List<DGVTemplateColumn> {
                new DGVTemplateColumn("Column1", "Column1 Name", 100, DataGridViewAutoSizeColumnMode.None, false),
                new DGVTemplateColumn("Column2", "Column2 Name", 200, DataGridViewAutoSizeColumnMode.None, true),
                new DGVTemplateColumn("Column3", "Column3 Name", 200, DataGridViewAutoSizeColumnMode.Fill, true),
                new DGVTemplateColumn("Column4", "Column4 Name", 300, DataGridViewAutoSizeColumnMode.NotSet, true)
            };
            var T = new DGVTemplate { Columns = TC };
            SetTemplate(T);
        }

        private void ReCreateDGV()
        {
            SkipDGVEvent = true;
            DGV_Template.Columns.Clear();
            Design.Columns.ForEach(C => DGV_Template.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = C.Name, Tag = C }));

            SkipDGVEvent = false;
            RefreshDGV();
        }

        private void RefreshDGV()
        {
            SkipDGVEvent = true;

            foreach (DataGridViewTextBoxColumn C in DGV_Template.Columns)
            {
                var DC = (DesignColumn)C.Tag;
                C.HeaderText = DC.Header;
                C.AutoSizeMode = DC.AutoSizeMode;
                C.Width = DC.Width;
                C.DataPropertyName = DC.Name;
                C.SortMode = DataGridViewColumnSortMode.NotSortable;
                C.DefaultCellStyle.ForeColor = DC.ForeColor;
                C.DefaultCellStyle.BackColor = DC.Visible ? Color.White : Color.SlateGray;
                //C.DefaultCellStyle.Font = C.Font;
            }

            var DT = new DataTable();
            Design.Columns.ForEach(C => DT.Columns.Add(new DataColumn(C.Name)));
            var R = DT.NewRow();
            Design.Columns.ForEach(C => R[C.Name] = $"{C.Name} Data");
            DT.Rows.Add(R);
            DGV_Template.DataSource = DT;

            if (BS_Columns.Position > 0)
            {
                DGV_Template.ClearSelection();
                DGV_Template[BS_Columns.Position, 0].Selected = true;
            }

            SkipDGVEvent = false;
        }

        #region BS Events

        private void BS_Columns_CurrentChanged(object sender, EventArgs e)
        {
            PG_Column.SelectedObject = BS_Columns.Current;
        }

        private void BS_Columns_CurrentItemChanged(object sender, EventArgs e)
        {
            RefreshDGV();
        }

        private void BS_Columns_ListChanged(object sender, ListChangedEventArgs e)
        {
            if (e.ListChangedType == ListChangedType.ItemDeleted || e.ListChangedType == ListChangedType.ItemAdded)
            {
                ReCreateDGV();
            }
        }

        #endregion BS Events

        #region DGV Events

        private void DGV_Template_CellMouseClick(object sender, DataGridViewCellMouseEventArgs e)
        {
            if (e.ColumnIndex >= 0) { BS_Columns.Position = e.ColumnIndex; }
        }

        private void DGV_Template_CellMouseDown(object sender, DataGridViewCellMouseEventArgs e)
        {
            Debug.Print($"Down: {e.Button}");
        }

        private void DGV_Template_CellMouseUp(object sender, DataGridViewCellMouseEventArgs e)
        {
            Debug.Print($"Up: {e.Button}");
            if (e.Button == MouseButtons.Left) { CheckOrder(); }
        }

        private void DGV_Template_ColumnDisplayIndexChanged(object sender, DataGridViewColumnEventArgs e)
        {
            Debug.Print($"Order I:{e.Column.HeaderText} DI:{e.Column.DisplayIndex}");
        }

        private void DGV_Template_ColumnHeaderMouseClick(object sender, DataGridViewCellMouseEventArgs e)
        {
            if (e.ColumnIndex >= 0) { BS_Columns.Position = e.ColumnIndex; }
        }

        #endregion DGV Events

        private void TSMI_Reset_Click(object sender, EventArgs e) => PG_Column.ResetSelectedProperty();
    }
}